codex-rs/core/src/chat_completions.rs

```diff
@@ -462,3 +465,228 @@ pub(crate) trait AggregateStreamExt: Stream<Item = Result<ResponseEvent>> + Size
+        };
+
+        let events = vec![
+            Ok(text_chunk("foo")),
+            Ok(ResponseEvent::OutputItemDone(func_call.clone())),
+            Ok(text_chunk("bar")),
+            Ok(ResponseEvent::Completed {
+                response_id: "r2".to_string(),
+                token_usage: None,
+            }),
+        ];
+
+        let stream = stream::iter(events).aggregate();
+        let collected: Vec<_> = stream.map(Result::unwrap).collect().await;
+
+        assert_eq!(collected.len(), 3);
+
+        // First event should be the function call forwarded directly.
+        assert!(matches!(
+            collected[0],
+            ResponseEvent::OutputItemDone(ResponseItem::FunctionCall { .. })
+        ));
+
+        // Second is the combined assistant message.
+        match &collected[1] {
+            ResponseEvent::OutputItemDone(ResponseItem::Message { content, .. }) => {
+                let text = match &content[0] {
+                    ContentItem::OutputText { text } => text,
+                    _ => panic!("unexpected content item"),
+                };
+                assert_eq!(text, "foobar");
+            }
+            other => panic!("unexpected second event: {other:?}"),
+        }
+
+        // Final Completed event.
+        assert!(matches!(
+            collected[2],
+            ResponseEvent::Completed { response_id: ref id, token_usage: None } if id == "r2"
+        ));
+    }
+
+    #[tokio::test]
+    async fn formats_tool_calls_in_chat_payload() {
+        use serde_json::Value;
+        use std::sync::Arc;
+        use std::sync::Mutex;
+        use wiremock::Mock;
+        use wiremock::MockServer;
+        use wiremock::Request;
+        use wiremock::Respond;
+        use wiremock::ResponseTemplate;
+        use wiremock::matchers::method;
+        use wiremock::matchers::path;
+
+        struct CaptureResponder(Arc<Mutex<Option<Value>>>);
+        impl Respond for CaptureResponder {
+            fn respond(&self, req: &Request) -> ResponseTemplate {
+                let v: Value = serde_json::from_slice(&req.body).unwrap();
+                *self.0.lock().unwrap() = Some(v);
+                ResponseTemplate::new(200)
+                    .insert_header("content-type", "text/event-stream")
+                    .set_body_raw(
+                        "event: response.completed\n\
+data: {\"type\":\"response.completed\",\"response\":{\"id\":\"resp\",\"output\":[]}}\n\n",
+                        "text/event-stream",
+                    )
+            }
+        }
+
+        let server = MockServer::start().await;
+        let captured = Arc::new(Mutex::new(None));
+
+        Mock::given(method("POST"))
+            .and(path("/v1/chat/completions"))
+            .respond_with(CaptureResponder(captured.clone()))
+            .expect(1)
+            .mount(&server)
+            .await;
+
+        unsafe {
+            std::env::set_var("OPENAI_REQUEST_MAX_RETRIES", "0");
```

Comment:
At some point, we should really find another way to thread this through so we can eliminate all these `unsafe` blocks.